src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1864,6 +1880,30 @@ private TypeParameterConstraintClauseSyntax ParseTypeParameterConstraintClause()
             {
                 _pool.Free(bounds);
             }
+
+            PostSkipAction skipBadTypeParameterConstraintTokens(SeparatedSyntaxListBuilder<TypeParameterConstraintSyntax> list, SyntaxKind expected)
+            {
+                CSharpSyntaxNode tmp = null;
+                Debug.Assert(list.Count > 0);
+                return this.SkipBadSeparatedListTokensWithExpectedKind(ref tmp, list,
+                    s_notExpectedForEndOfTypeConstraintClause,
+                    s_abortForEndOfTypeConstraintClause,
+                    expected);
+            }
+        }
+
+        private static readonly Func<LanguageParser, bool> s_notExpectedForEndOfTypeConstraintClause = p => p.NotExpectedForEndOfTypeConstraintClause();
+        private bool NotExpectedForEndOfTypeConstraintClause()
+        {
+            return this.CurrentToken.Kind != SyntaxKind.CommaToken && !this.IsPossibleTypeParameterConstraint();
+        }
+
+        private static readonly Func<LanguageParser, bool> s_abortForEndOfTypeConstraintClause = p => p.AbortForEndOfTypeConstraintClause();
+        private bool AbortForEndOfTypeConstraintClause()
+        {
+            return this.CurrentToken.ContextualKind == SyntaxKind.WhereKeyword
+                   || this.CurrentToken.Kind == SyntaxKind.EqualsGreaterThanToken // ex: 'void localFunction<T>() where T : class => ...'
```

Comment:
Note: we could probably handle the `=>` as a terminator as well, but that seems out of scope for current PR.